src/neo/SmartContract/ApplicationEngine.Runtime.cs

```diff
@@ -228,6 +234,20 @@ protected internal int GetInvocationCounter()
             return counter;
         }
 
+        /// <summary>
+        /// The implementation of System.Runtime.GetRandom.
+        /// Gets the random number genrated form the VRF
+        /// Primary geterates this random number with `prevHash`, the hash of the previous (validators.Length/3 + 1) block
+        /// </summary>
+        /// <returns>The last four bytes of the random number.</returns>
+        protected internal uint GetRandom()
+        {
+            if (nextNonce == null) throw new InvalidOperationException("Require persistingBlock");
+            var nonce = nextNonce;
+            nextNonce = Cryptography.Helper.Sha256(nonce);
```

Comment:
@Liaojinghui If I know the `nonce`, I can compute the next value before call the random method

Reply:
> @Liaojinghui If I know the `nonce`, I can compute the next value before call the random method

The problem is the `nonce` generates after transactions are sent but before they are executed. When you see the `nonce`, it has already been used. The nonce in the header is not for the next block, but for the current one.

Reply:
I also did not understand the attack vector you are mentioning @shargon.

I see if the first transaction write in the storage and the second one reads it. Then, I see that nonce could be shared.
But then the problem would be on the Smart Contract itself, in my perspective, which would allow read data to be used.

Reply:
`nonce` will be exposed by `ToStackItem` ?

Reply:
> When you see the nonce, it has already been used.

It depends, the primary see the nonce before it, and the next primary can compute the nonce with the prepare request of the previous block.

Reply:
> > When you see the nonce, it has already been used.
> 
> It depends, the primary see the nonce before it, and the next primary can compute the nonce with the prepare request of the previous block.

You and @roman-khimov are concerning on the same issue. But I consider `nonce` as a tool, and developers can decide how to use it.  I am also working on `automated smart contract`, with it, we can feed the `nonce` in a callback transaction.